template/context.go

```diff
@@ -58,7 +67,166 @@ func NewContextFromTask(t *api.Task) (ctx Context) {
+		file := secretRef.GetFile()
+		if file != nil && file.Name == target {
+			secret := ctx.restrictedSecrets.Get(secretRef.SecretID)
+			if secret != nil {
+				return string(secret.Spec.Data), nil
+			}
+			break
+		}
+	}
+
+	return "", errors.Errorf("secret target %s not found", target)
+}
+
+func (ctx PayloadContext) configGetterBySource(source string) (string, error) {
+	if ctx.restrictedConfigs == nil {
+		return "", errors.New("configs unavailable")
+	}
+
+	container := ctx.t.Spec.GetContainer()
+	if container == nil {
+		return "", errors.New("task is not a container")
+	}
+
+	for _, configRef := range container.Configs {
+		if configRef.ConfigName == source {
+			config := ctx.restrictedConfigs.Get(configRef.ConfigID)
+			if config != nil {
+				return string(config.Spec.Data), nil
+			}
+			break
+		}
+	}
+
+	return "", errors.Errorf("config source %s not found", source)
+}
+
+func (ctx PayloadContext) configGetterByTarget(target string) (string, error) {
+	if ctx.restrictedConfigs == nil {
+		return "", errors.New("configs unvailable")
+	}
+
+	container := ctx.t.Spec.GetContainer()
+	if container == nil {
+		return "", errors.New("task is not a container")
+	}
+
+	for _, configRef := range container.Configs {
+		file := configRef.GetFile()
+		if file != nil && file.Name == target {
+			config := ctx.restrictedConfigs.Get(configRef.ConfigID)
+			if config != nil {
+				return string(config.Spec.Data), nil
+			}
+			break
+		}
+	}
+
+	return "", errors.Errorf("config target %s not found", target)
+}
+
+func (ctx PayloadContext) envGetter(variable string) (string, error) {
+	container := ctx.t.Spec.GetContainer()
+	if container == nil {
+		return "", errors.New("task is not a container")
+	}
+
+	for _, env := range container.Env {
+		parts := strings.SplitN(env, "=", 2)
+
+		if len(parts) > 1 && parts[0] == variable {
+			return parts[1], nil
+		}
+	}
+	return "", nil
+}
+
+// NewPayloadContextFromTask returns a new template context from the data
+// available in the task. This context also provides access to the configs
+// and secrets that the task has access to. The provided context can then
+// be used to populate runtime values in a templated config or secret.
+func NewPayloadContextFromTask(t *api.Task, dependencies exec.DependencyGetter) (ctx PayloadContext) {
+	return PayloadContext{
+		Context:           NewContextFromTask(t),
+		t:                 t,
+		restrictedSecrets: secrets.Restrict(dependencies.Secrets(), t),
+		restrictedConfigs: configs.Restrict(dependencies.Configs(), t),
+	}
+}
+
+// Expand treats the string s as a template and populates it with values from
+// the context.
+func (ctx *PayloadContext) Expand(s string) (string, error) {
+	funcMap := template.FuncMap{
+		"SecretBySource": ctx.secretGetterBySource,
+		"Secret":         ctx.secretGetterByTarget,
+		"ConfigBySource": ctx.configGetterBySource,
+		"Config":         ctx.configGetterByTarget,
+		"Env":            ctx.envGetter,
```

Comment:
The "best practice" seems to be that variables are uppercase and functions are lowercase. This was followed with `join`.

Here is the builtin function set: https://golang.org/pkg/text/template/#hdr-Functions.

I think we follow this convention in docker, as well: https://github.com/moby/moby/blob/master/pkg/templates/templates.go#L11.